From f3ea69a157e5018c47284f395c979ee575dea96a Mon Sep 17 00:00:00 2001
From: Anthony Lopez <anthlope@codeaurora.org>
Date: Wed, 10 Mar 2021 15:44:40 -0800
Subject: [PATCH 1/1] User-specified destination port patch

Add port number to the list of command line arguments. If provided, add this
port number to the destination IP parameters.

Signed-off-by: Anthony Lopez <anthlope@codeaurora.org>
---
 src/udpreplay.cpp | 39 +++++++++++++++++++++++++++++++++------
 1 file changed, 33 insertions(+), 6 deletions(-)

diff --git a/src/udpreplay.cpp b/src/udpreplay.cpp
index dd0e2c0..f90aee7 100644
--- a/src/udpreplay.cpp
+++ b/src/udpreplay.cpp
@@ -13,6 +13,7 @@
 #include <unistd.h>
 
 #define NANOSECONDS_PER_SECOND 1000000000L
+#define MAX_UDP_PORT_RANGE 65535
 
 int main(int argc, char *argv[]) {
 
@@ -25,9 +26,10 @@ int main(int argc, char *argv[]) {
   int broadcast = 0;
   std::string dest_ipv4 = "";
   std::string dest_ipv6 = "";
+  int dest_port = 0;
 
   int opt;
-  while ((opt = getopt(argc, argv, "i:bls:c:r:t:d:g:")) != -1) {
+  while ((opt = getopt(argc, argv, "i:bls:c:r:t:d:g:p:")) != -1) {
     switch (opt) {
     case 'i':
       ifindex = if_nametoindex(optarg);
@@ -77,6 +79,15 @@ int main(int argc, char *argv[]) {
       printf("Will attempt to send to ipv6 address: %s\n", optarg);
       dest_ipv6 = (optarg);
       break;
+    case 'p':
+      dest_port = std::stoi(optarg);
+      if ((dest_port <= 0) || (dest_port > MAX_UDP_PORT_RANGE)) {
+        std::cerr << "Invalid port number." << std::endl;
+        return 1;
+      }
+      std::cout << "Attempting to send to destination port :"
+                    << dest_port << std::endl;
+      break;
     default:
       goto usage;
     }
@@ -87,7 +98,7 @@ int main(int argc, char *argv[]) {
         << "udpreplay 1.0.0 © 2020 Erik Rigtorp <erik@rigtorp.se> "
            "https://github.com/rigtorp/udpreplay\n"
            "usage: udpreplay [-i iface] [-l] [-s speed] [-c millisec] [-r "
-           "repeat] [-t ttl] [-d dest ipv4] [-g dest ipv6] "
+           "repeat] [-t ttl] [-d dest ipv4] [-g dest ipv6] [-p destination port] "
            "pcap\n"
            "\n"
            "  -i iface    interface to send packets through\n"
@@ -98,7 +109,8 @@ int main(int argc, char *argv[]) {
            "  -t ttl      packet ttl\n"
            "  -b          enable broadcast (SO_BROADCAST)\n"
            "  -d          destination ipv4 address\n"
-           "  -g          destination ipv6 address"
+           "  -g          destination ipv6 address\n"
+           "  -p          destination port number"
         << std::endl;
     return 1;
   }
@@ -251,7 +263,12 @@ int main(int argc, char *argv[]) {
         struct sockaddr_in6 addr_ipv6;
         memset(&addr_ipv6, 0, sizeof(addr_ipv6));
         addr_ipv6.sin6_family = AF_INET6;
-        addr_ipv6.sin6_port = udp->uh_dport;
+        // check if destination port is provided
+        if(dest_port > 0) {
+          addr_ipv6.sin6_port = htons(dest_port);
+        } else {
+          addr_ipv6.sin6_port = udp->uh_dport;
+        }
         inet_pton(AF_INET6, dest_ipv6.data(), &(addr_ipv6.sin6_addr));
         auto n = sendto(fd, d, len, 0,
                           reinterpret_cast<sockaddr *>(&addr_ipv6),
@@ -266,9 +283,19 @@ int main(int argc, char *argv[]) {
         addr.sin_family = AF_INET;
 
 #ifdef __GLIBC__
-        addr.sin_port = udp->dest;
+        // check if destination port is provided
+        if(dest_port > 0) {
+          addr.sin_port = htons(dest_port);
+        } else {
+          addr.sin_port = udp->dest;
+        }
 #else
-        addr.sin_port = udp->uh_dport;
+        // check if destination port is provided
+        if(dest_port > 0) {
+          addr.sin_port = htons(dest_port);
+        } else {
+          addr.sin_port = udp->uh_dport;
+        }
 #endif
 
         // use provided ipv4 destination address
-- 
2.29.0

